crates/iceberg/src/metadata_scan.rs

```diff
@@ -0,0 +1,265 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Metadata table api.
+
+use std::sync::Arc;
+
+use arrow_array::builder::{MapBuilder, PrimitiveBuilder, StringBuilder};
+use arrow_array::types::{Int64Type, TimestampMillisecondType};
+use arrow_array::RecordBatch;
+use arrow_schema::{DataType, Field, Schema, TimeUnit};
+
+use crate::spec::TableMetadataRef;
+use crate::table::Table;
+use crate::Result;
+
+/// Table metadata scan.
+///
+/// Used to inspect a table's history, snapshots, and other metadata as a table.
+///
+/// See also <https://iceberg.apache.org/docs/latest/spark-queries/#inspecting-tables>.
+#[derive(Debug)]
+pub struct MetadataScan {
+    metadata_ref: TableMetadataRef,
+}
+
+impl MetadataScan {
+    /// Creates a new metadata scan.
+    pub fn new(table: &Table) -> Self {
+        Self {
+            metadata_ref: table.metadata_ref(),
+        }
+    }
+
+    /// Returns the snapshots of the table.
+    pub fn snapshots(&self) -> Result<RecordBatch> {
+        SnapshotsTable::scan(self)
+    }
+}
+
+/// Table metadata scan.
+///
+/// Use to inspect a table's history, snapshots, and other metadata as a table.
+///
+/// References:
+/// - <https://github.com/apache/iceberg/blob/ac865e334e143dfd9e33011d8cf710b46d91f1e5/core/src/main/java/org/apache/iceberg/MetadataTableType.java#L23-L39>
+/// - <https://iceberg.apache.org/docs/latest/spark-queries/#querying-with-sql>
+/// - <https://py.iceberg.apache.org/api/#inspecting-tables>
+pub trait MetadataTable {
```

Comment:
Hi, I'm a bit confused about why we need this trait. Could you show how we would use it? Is there a use case that we need `impl MetadataTable` or `dyn MetadataTable`? Does it make sense to provide API like `table.snapshot_table()`?

Reply:
Yes, it's not a must-have. Just an common interface for different metadata tables.

One practical motivation is that I want to ensure `schema` method exists, like `SnapshotsTable::schema()`.

Reply:
If so, directly constructing structs makes more sense to me, as it makes the API more intuitive to use.

Would you like to give it a try?

Reply:
Previously I thought the schema is always a constant, so I designed the interface so that query engine can get the schema without loading the table from catalog. 

But it seems not the case. As @rshkv mentioned https://github.com/apache/iceberg-rust/pull/822#discussion_r1898945557

e.g., `readable_metrics` is dynamic https://py.iceberg.apache.org/api/#entries


Therefore, what about this alternative design:
```rust
// user API
table.metadata_scan().snapshots().schema()
table.metadata_scan().snapshots().scan()

// impl

impl MetadataScan {
    /// Returns the snapshots of the table.
    pub fn snapshots(&self) -> Result<RecordBatch> {
        SnapshotsTable{ metadata: self }
    }
}

struct SnapshotsTable<'a> {
    metadata: &'a MetadataScan,
}

// trait is just to ensure interface, and add doc
pub trait MetadataTable {
    /// Returns the schema of the metadata table.
    fn schema(&self) -> Schema;

    /// Scans the metadata table.
    fn scan(&self) -> Result<RecordBatch>;
}
```

Reply:
Oh, didn't see you comment above. I think whether or not to have the trait are both OK to me

Reply:
updated

Reply:
:+1: removing this trait definition looks good to me